pgms/volumeselect/volumeselect.man
.ds Vi Vista
.ds Vn 2.0
.TH volumesel 1Vi "15 Februar 1994" "\*(Vi Version \*(Vn"
.SH NAME
volumesel \- select volumes from a volume set
.SH SYNOPSIS
\fBvolumesel\fR \ <options>
[\fIinfile\fR] [\fIoutfile\fR]
.SH DESCRIPTION
\fBvolumesel\fP selects volumes from a set of volumes according
to some given selection criterion. Possible criteria are:
size, circularity and compactness.

"Size" refers to be number of voxels belonging to the volume.

"Circularity" is a measure of roundness defined as:

  (3.0 * size) / (4.0 * pi * rmax)

where "rmax" is the radius of the smallest enclosing sphere.
Thus if the volume is a sphere, this value will be "1",
The further the volume's shape deviates from spherical
shape, the smaller this value will become.


"Compactness" is defined as the ratio between the number of border voxels
and volume size. Thus, if most voxels belonging to the volume are
interior voxels this value will be close to "1".

The input image must be of type "volumes". The output image
will be also be of type "volumes".

.SH COMMAND LINE OPTIONS
\fBvolumesel\fP accepts the following options:
.IP \fB-help\fP 15n
Prints a message describing options.
.IP "\fB-in\fP \fIinfile\fP"
Specifies the Vista data file containing the first input image.
.IP "\fB-image\fP \fIinfile\fP"
Specifies the Vista data file containing the second input image.
.IP "\fB-out\fP \fIoutfile\fP"
Specifies where to write the output images as a Vista data file.
.IP "\fB-feature\fP \fIoutfile\fP"
Specifies the selection criterion. Possible choices are:
size | circularity | compactness. Default: size.
.IP "\fB-min\fP \fImin\fP"
Specifies the lower bound for the feature value.
.IP "\fB-max\fP \fImax\fP"
Specifies the upper bound for the feature value.
.LP
Input and output files can be specified on the command line or allowed to
default to the standard input and output streams.
.SH "SEE ALSO"
.na
.nh
.BR VImage (3Vi),
.BR Vista (7Vi)
.hy
.ad
.SH AUTHOR
Gabriele Lohmann <lohmann@cns.mpg.de>


debian/changelog
via (2.0.1svn191-0) stable; urgency=low

  *renamed via-bin to via-utils

 -- Lipsia Group (The official Lipsia developer group) <lipsia@cbs.mpg.de>  Tue, 09 Nov 2010 12:17:00 +0400 

via (2.0.0-0) stable; urgency=low

  *This version comes with the new lipsia-2.0 package
   and comprises some minor bug fixes
  *Fixed type error of VLong (is now long)

 -- Lipsia Group (The official Lipsia developer group) <lipsia@cbs.mpg.de>  Tue, 09 Nov 2010 12:17:00 +0400 

via (1.6.0-2) unstable; urgency=low

  * New internal release

 -- Frank Burkhardt <burk@cbs.mpg.de>  Mon, 20 Jul 2009 11:45:41 +0200

via (1.6.0-1) unstable; urgency=low

  * Initial release (Closes: #521328).
  * New Upstream Version. Renamed 'vcat' to 'vcatobj' to not conflict with
    atfs package in Debian.
  * Update Standards-version to 3.8.2.0 -- no changes necessary.
  * Update watchfile for recent change in upstream naming scheme.
  * Add versioned dependency on debhelper (>= 5) upon recommendation by
    lintian.

 -- Michael Hanke <michael.hanke@gmail.com>  Fri, 19 Jun 2009 13:50:52 +0200

via (1.5.2+svn3529-2) UNRELEASED; urgency=low

  * Add alternative build-dependency on atlas3-base-dev to be able to build
    the package on Debian etch.
  * FTP access to the upstream sources needs passive mode. Fixed debian/watch.

 -- Michael Hanke <michael.hanke@gmail.com>  Tue, 14 Apr 2009 02:46:10 +0200

via (1.5.2+svn3529-1) UNRELEASED; urgency=low

  * New Upstream SVN snapshot that relocates the VIA lib headers to their new
    cannonical location.

 -- Michael Hanke <michael.hanke@gmail.com>  Tue, 31 Mar 2009 22:08:01 +0200

via (1.5.2-1) UNRELEASED; urgency=low

  * Initial Debian packaging. This is a complete rewrite of the original
    Debian packaging effort done by the upstream developers.

 -- Michael Hanke <michael.hanke@gmail.com>  Thu, 26 Mar 2009 19:46:09 +0100


pgms/volumeselect/volumeselect.c
/****************************************************************
 *
 * Copyright (C) Max Planck Institute 
 * for Human Cognitive and Brain Sciences, Leipzig
 *
 * Author Gabriele Lohmann, 2004, <lipsia@cbs.mpg.de>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 *
 * $Id: volumeselect.c 3177 2008-04-01 14:47:24Z karstenm $
 *
 *****************************************************************/

/*! \brief volumeselect - select volumes from a set of volumes


\par Description
Select volumes from a set of volumes.
The output image will contain the selected volumes.


\par Usage

        <code>volumeselect</code>

        \param -in      input volume set
        \param -out     output volume set
        \param -feature type of feature (size | circularity | compactness). Default: size
        \param -type    type of operation (range | min | max). Default: range
        \param -min     min value. Default: 0
        \param -max     max value. Default: 1


\par Examples
<br>
\par Known bugs
none.

\file volumeselect.c
\author G.Lohmann, MPI-CBS, 2004
*/



/* From the Vista library: */
#include <viaio/Vlib.h>
#include <viaio/mu.h>
#include <viaio/option.h>
#include <via.h>

/* From the standard C libaray: */
#include <stdio.h>
#include <math.h>
#include <stdlib.h>


enum FEATURE {SIZE, CIRCULARITY, COMPACTNESS};

VDictEntry FEATUREDict[] = {
  { "size", 0 },
  { "circularity", 1 },
  { "compactness", 2 },
  { NULL }
};


VDictEntry TypeDict[] = {
  { "range", 0 },
  { "min", 1 },
  { "max", 2 },
  { NULL }
};


extern Volumes VolumeSelect(Volumes,VLong,VLong,double,double);

int 
main (int argc, char *argv[])
{
  static VLong   feature = 0;
  static VLong   type    = 0;
  static VDouble minval = 0;
  static VDouble maxval = 1.0;
  static VOptionDescRec options[] = {
    { "feature", VLongRepn, 1,(VPointer) & feature, 
      VOptionalOpt, FEATUREDict,"type of feature" },
    { "type", VLongRepn, 1,(VPointer) & type, 
      VOptionalOpt, TypeDict,"type of operation: range, min, max" },
    { "min", VDoubleRepn, 1,(VPointer) & minval, 
      VOptionalOpt, NULL,"min value" },
    { "max", VDoubleRepn, 1,(VPointer) & maxval, 
      VOptionalOpt, NULL,"max value" }
  };
  FILE *in_file,*out_file;
  VAttrList list;
  Volumes src=NULL,dest=NULL;
  VImage src_image=NULL,dst_image=NULL,tmp=NULL;
  VAttrListPosn posn;
  int nl=0;
  char prg[50];	
  sprintf(prg,"volumeselect V%s", getVersion());
  fprintf (stderr, "%s\n", prg);

  /* Parse command line arguments: */
  VParseFilterCmd (VNumber (options), options, argc, argv,&in_file,&out_file);


  /* Read source image(s): */
  if (! (list = VReadFile (in_file, NULL))) exit (1);

  /* Performs conversion on each image: */
  for (VFirstAttr (list, & posn); VAttrExists (& posn); VNextAttr (& posn)) {

    switch (VGetAttrRepn (& posn)) {

    case VImageRepn:

      VGetAttrValue (& posn, NULL, VImageRepn, & src_image);

      if (VPixelRepn(src_image) == VBitRepn) {
	tmp = VLabelImage3d(src_image,NULL,26,VShortRepn,&nl);
	src = VImage2Volumes(tmp);
      }
      else {
	src  = VImage2Volumes(src_image);
      }
      dest = VolumeSelect(src,feature,type,minval,maxval);
      if (dest == NULL) VError("no volumes selected");
      dst_image = Volumes2Image(dest,VPixelRepn(src_image));
      VSetAttrValue (& posn, NULL, VImageRepn, dst_image);
      break;

    case VolumesRepn:

      VGetAttrValue (& posn, NULL, VolumesRepn, & src);
      dest = VolumeSelect(src,feature,type,minval,maxval);
      if (dest == NULL) VError("no volumes selected");
      VSetAttrValue (& posn, NULL, VolumesRepn, dest);
      break;

    default:
      continue;
    }
  }

  /* Write the results to the output file: */
  VHistory(VNumber(options),options,prg,&list,&list);
  if (VWriteFile (out_file, list))
    fprintf (stderr, "%s: done.\n",argv[0]);
  return 0;
}


/*
** select volumes
*/
Volumes
VolumeSelect(Volumes src,VLong feature,VLong type,double minval,double maxval)
{
  Volumes dest;
  Volume u,v,v0;
  double size;
  double mean[3];
  double pi = 3.14159265;
  double rmax;
  double bordersize;
  double value = 0;
  int nvol,nsel;
  int nbands,nrows,ncols;
  double sum1, sum2, ave, sigma;
  VDouble xmin,xmax;

  /*
  ** create new Volume set
  */
  nbands = VolumesNBands(src);
  nrows  = VolumesNRows(src);
  ncols  = VolumesNColumns(src);
  dest   = VCreateVolumes(nbands,nrows,ncols);


  xmin = VRepnMaxValue(VDoubleRepn);
  xmax = VRepnMinValue(VDoubleRepn);

  sum1 = sum2 = 0;
  nsel = 0;
  nvol = 0;
  v0   = NULL;

  for (v = src->first; v != NULL; v = v->next) {

    switch (feature) {

    case SIZE:

      value = (double) VolumeSize(v);
      break;

    case CIRCULARITY:

      size = (double) VolumeSize(v);
      VolumeCentroid(v,&mean[0]);
      rmax = VolumeRadius(v,&mean[0]);
      rmax = rmax * rmax * rmax;
      value = (double) 3.0 * size / ((double) 4.0 * pi * rmax);
      break;

    case COMPACTNESS:
      
      size = (double) VolumeSize(v);
      bordersize = VolumeBorderSize(v);
      value = bordersize / size;
      break;

    default:
      VError("Illegal choice of feature");
    }


    if (type == 0 && value >= minval && value <= maxval) {
      u = VCopyVolume(v);
      VAddVolume(dest,u);
      nsel++;
    }

    if (value < xmin) {
      xmin = value;
      if (type == 1) {
	v0 = v;
	nsel = 1;
      }
    }

    if (value > xmax) {
      xmax = value;
      fprintf(stderr," xmax= %f\n",xmax);
      if (type == 2) {
	v0 = v;
	nsel = 1;
      }
    }

    sum1 += value * value;
    sum2 += value;

    nvol++;
  }

  if (nsel == 0) return NULL;

  if (v0 != NULL) {
    u = VCopyVolume(v0);
    VAddVolume(dest,u);
  }

  ave   = sum2 / (double) nvol;
  sigma = 0;
  if (nvol > 1)
    sigma = sqrt((sum1 - sum2 * sum2 / (double) nvol) / ((double) nvol - 1.0));
  
  fprintf(stderr,"  min: %12.4f,    max: %12.4f\n",xmin,xmax);
  fprintf(stderr," mean: %12.4f,  sigma: %12.4f\n",ave,sigma);
  fprintf(stderr,"%d out of %d volumes selected \n",nsel,nvol);
  return dest;
}
